cpp/src/tsne/tsne.cu

```diff
@@ -0,0 +1,241 @@
+    if (verbose) printf("[Info] Now normalizing distances so exp(D) doesn't explode.\n");
+    normalize_distances(n, distances, n_neighbors, stream);
+
+
+    // Optimal perplexity
+    if (verbose) printf("[Info] Searching for optimal perplexity via bisection search.\n");
+    float *P = (float *)d_alloc->allocate(sizeof(float) * n * n_neighbors, stream);
+
+    // Determine best blocksize / gridsize
+    int blockSize_N = 1024; int minGridSize_N;
+    cuda_max_potential(&minGridSize_N, &blockSize_N, __determine_sigmas, 0, n);
+    const int gridSize_N = ceil(n, blockSize_N);
+
+
+    const float P_sum = determine_sigmas(distances, P, perplexity, perplexity_max_iter,
+                                        perplexity_tol, n, n_neighbors, stream, gridSize_N, blockSize_N);
+    d_alloc->deallocate(distances, sizeof(float) * n * n_neighbors, stream);
+    if (verbose) printf("[Info] Perplexity sum = %f\n", P_sum);
+
+
+    // Convert data to COO layout
+    COO_t<float> P_PT;
+    symmetrize_perplexity(P, indices, &P_PT, n, n_neighbors, P_sum, early_exaggeration, stream, handle);
+        
+    const int NNZ = P_PT.nnz;
+    float *VAL = P_PT.vals;
+    const int *COL = P_PT.rows;
+    const int *ROW = P_PT.cols;
+
+
+    // Allocate data [NOTICE Fortran Contiguous for method = Naive and C-Contiguous for fast]
+    if (initialize_embeddings)
+        random_vector(Y, -0.03f, 0.03f, n * k, stream, seed);
+
+
+    // Allocate space
+    if (verbose) printf("[Info] Now allocating memory for TSNE.\n");
+    float *norm = (float *)d_alloc->allocate(sizeof(float) * n, stream);
+    float *Q_sum = (float *)d_alloc->allocate(sizeof(float) * n, stream);
+    double *sum = (double *)d_alloc->allocate(sizeof(double), stream);
+
+    float *attract = (float *)d_alloc->allocate(sizeof(float) * n * k, stream);
+    float *repel = (float *)d_alloc->allocate(sizeof(float) * n * k, stream);
+
+    float *iY = (float *)d_alloc->allocate(sizeof(float) * n * k, stream);
+    float *gains = (float *)d_alloc->allocate(sizeof(float) * n * k, stream);
+    float *means = (float*)d_alloc->allocate(sizeof(float) * k, stream);
+
+
+    // Compute optimal gridSize and blockSize for attractive forces
+    int blockSize_NNZ = 1024; int minGridSize_NNZ;
+    if (n_components == 2)
+        cuda_max_potential(&minGridSize_NNZ, &blockSize_NNZ, __attractive_fast_2dim, 0, NNZ);
+    else
+        cuda_max_potential(&minGridSize_NNZ, &blockSize_NNZ, __attractive_fast, 0, NNZ);
+    const int gridSize_NNZ = ceil(NNZ, blockSize_NNZ);
+
+    // Compute optimal gridSize and blockSize for applying forces
+    int blockSize_dimN = 1024; int minGridSize_dimN;
+    cuda_max_potential(&minGridSize_dimN, &blockSize_dimN, __apply_forces, 0, n_components*n);
+    const int gridSize_dimN = ceil(n_components*n, blockSize_dimN);
+
+
+    // Do gradient updates
+    float momentum = pre_momentum;
+    float Z;
+
+    if (verbose) printf("[Info] Start gradient updates!\n");
+    if (method == 1) {
```

Comment:
For readability, let's pull each of these methods into separate functions within `tsne.cu` so that we can just have:

```
switch(TSNEMethod) {
    case TSNEMethod::Naive:
        perform_naive();
        break;
    case TSNEMethod::Fast:
        perform_fast();
}
```